### ISSUE TITLE
Small Tutorial for Compiling on Raspberry Pi

### ISSUE THREAD

**Issue by Nexxurs** [NagiosEnterprises/ncpa#477]
Hey!

I'm not sure where to put this, so here it is :P

Yesterday i tried to compile NCPA for the Raspberry Pi for about 4 hours and had quite a number of hurdles to take. 
I want to make this easier for anyone else trying it, so here are some insights:

(I'm on a train right now and dont have the correct commands with me, ill try to do it as best as I can from memory - if you like I can write a more thorough guide when I get home)

First of all, the default python versions won't work. Python 2.7.14 is packed within the repo as of now, but is absolutly unusable on the Raspberry Pi and Python2 in the Raspbian Repo is Version 2.7.13 (which results in file-not-found errors while compiling to rpm - took me a really long time to resolve that one)

So you have to compile your own Version of Python 2.7.14.
Download the Source from [Python.org](https://www.python.org/ftp/python/2.7.14/Python-2.7.14.tar.xz) and unpack it. Within the unpacked-directory run "make altinstall" to create a new python file without overwriting your own. 
(I believe that worked without any errors on my device - at least i dont recall anything grave)

You also need pip for your new Version of python, so download the current get-pip.py and run it with your new local python executable (./python, not python!)

After you compiled your own Version of Python 2.7.14 the file "python" is in the current directory. 
The Makefile of NCPA uses the /usr/local/bin/python2.7 installation of your device, so you need to create a symlink to it
If you already have a python link in there (check with /usr/local/bin/python2.7 --version), you can rename it so you can undo those changes afterwards. Run 
`mv /usr/local/bin/python2.7 /usr/local/bin/python2.7-backup`

Now you can create the symlink with 
`ln -s /full/path/to/new/python /usr/local/bin/python2.7`

You also need to install the other requirements. Officially they are:
gcc, gcc-c++
zlib, zlib-devel
openssl, openssl-devel
sqlite3, sqlite3-devel
libffi-devel
rpm-build
wget
git

But for the Raspberry they are a little bit different.
gcc-c++:	I didn't find that package, but gcc alone is enough
zlib		I don't believe i have it installed, but it didnt't matter
openssl-devel	Doesn't exist, but it worked anyways
sqlite3-devel	Use sqlite3-dev instead
libffi-devel	Use libffi-dev instead
rpm-build	Doesn't exist as Standalone - use rpm
wget		Should be installed anyways

alien		Is not in the requirements, but is still needed
libssl-dev	The linux_prereqs doesn't show it, but tries to install it if you use debian - you need it aswell

So in the end, you need 
`apt-get install gcc openssl sqlite3 sqlite3-dev libffi-dev rpm git debian-builder alien libssl-dev`
(I'm not sure if thats all - I installed a lot of stuff while trying to get it working. When I write the better Version, ill track exactly what i needed)

Download the repo and install the python requirements

```
git clone https://github.com/NagiosEnterprises/ncpa
cd ncpa/build
/usr/local/bin/python2.7 -m pip install -r resources/require.txt (probably sudo needed)
```

The next step is to modify the Makefile. If you don't specify the target architecture, rpm will choose some armv7XX, which cant be converted to deb by alien. 
On line 67 change 
`QA_RPATHS='$[ 0x0002 ]' rpmbuild $(BUILDROOT)/SPECS/ncpa.spec -bb --define '_topdir /usr/src/redhat'`
to 
`QA_RPATHS='$[ 0x0002 ]' rpmbuild $(BUILDROOT)/SPECS/ncpa.spec -bb --target=armhf --define '_topdir /usr/src/redhat'`

Now to the compiling itself:
`sudo make build_deb `
(it creates a buildpath into /usr/src, so it needs to write there)

Thats it. Now there should be a folder named debbuild with your debian file.

**Comment by jomann09** [comment]
Thanks for this, I will break it out into the docs that we have when I get a chance.

**Comment by jomann09** [comment]
Thank you for an update on this! I will have to test this out.

**Comment by neatchee** [comment]
Spoke too soon :(
I'm getting "ImportError: no module named os" when actually going to run the listener.

Looking elsewhere online this appears to be a cx_Freeze problem when it goes to import the python base.

I had to go back to 2.1.5 for now.

**Comment by jomann09** [comment]
I believe the reason you were having issues with this build was that you were using the newest version of cx_Freeze instead of the patched 4.3.4 that comes in the NCPA repo. I am working on using the latest cx_Freeze for NCPA 3 but for now we are using the older one. That being said, I had no issues building and running a deb package after editing up the build process a bit to make it work properly on Raspbian. I'm going to close this since the master branch has the proper build process and I am going to be building a Raspbian build from now on.
